src/validations/rules/ssp.sch

```diff
@@ -3813,6 +3819,102 @@
 
     </sch:pattern>
 
+    <sch:pattern
+        id="dns">
+
+        <sch:rule
+            context="oscal:system-implementation">
+
+            <sch:assert
+                diagnostics="has-DNS-authoritative-service-diagnostic"
+                id="has-DNS-authoritative-service"
+                role="information"
+                test="exists(oscal:component[@type eq 'DNS-authoritative-service'])">A DNS authoritative service is defined.</sch:assert>
+
+        </sch:rule>
+
+        <sch:rule
+            context="oscal:component[@type eq 'DNS-authoritative-service']">
+
+            <sch:assert
+                diagnostics="has-DNS-zone-diagnostic"
+                id="has-DNS-zone"
+                role="error"
+                test="exists(oscal:prop[@name eq 'DNS-zone' and exists(@value)])">The DNS authoritative service has one or more zones
+                specified.</sch:assert>
+
+        </sch:rule>
+
+        <sch:rule
+            context="oscal:component[@type eq 'DNS-authoritative-service' and oscal:status/@state eq 'operational']/oscal:prop[@name eq 'DNS-zone'][$use-remote-resources]">
+
+            <sch:let
+                name="zone-regex"
+                value="'^ ([a-z0-9]+ (-[a-z0-9]+)*\.)+ [a-z]{2,}\.?$'" />
+            <sch:let
+                name="well-formed"
+                value="matches(@value, $zone-regex, 'ix')" />
+            <sch:assert
+                diagnostics="has-well-formed-DNS-zone-diagnostic"
+                id="has-well-formed-DNS-zone"
+                role="error"
+                test="$well-formed">Each zone name for the DNS authoritative service is well-formed.</sch:assert>
+
+            <!-- ensure zone name used for query ends with dot -->
+            <sch:let
+                name="DoH_target"
+                value="
+                    if (ends-with(@value, '.')) then
+                        @value
+                    else
+                        concat(@value, '.')" />
+
+            <!-- See https://developers.google.com/speed/public-dns/docs/doh -->
+
+            <sch:let
+                name="DoH_query"
+                value="concat('https://dns.google/resolve?name=', $DoH_target, '&amp;type=SOA')" />
+
+            <sch:let
+                name="DoH_response"
+                value="
+                    if ($well-formed) then
+                        unparsed-text($DoH_query)
+                    else
+                        ()" />
+
+            <sch:let
+                name="response"
+                value="
+                    if ($well-formed) then
+                        parse-json($DoH_response)
+                    else
+                        ()" />
+
+            <sch:let
+                name="has-resolved-zone"
+                value="
+                    if ($well-formed) then
+                        $response?Status eq 0 and map:contains($response, 'Answer')
+                    else
+                        false()" />
```

Comment:
This is not thought out, but I wonder if these sorts of network check - out-of-process things that might fail - should have a separate assertion for network errors.

That would allow us to distinguish between an unmet requirement and an unrecoverable validation error.

(This comes to mind, because I'm not sure if Google sets CORS headers such that we can make this check in-browser.)

Reply:
Or would the diagnostic messages, below, serve that purpose? Wondering if such diagnostic messages should be added to the UI.

Reply:
We should probably choose a uniform manner with which remote resource retrieval failures are handled. I have created #469 to attend to that.